src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -450,191 +565,228 @@ public SubSourcePromiseControl(SourceShape<(
+    /// </remarks>
+    internal sealed class SubSourceStage<K, V, TMessage> : GraphStage<SourceShape<TMessage>>
+    {
+        private readonly TopicPartition _topicPartition;
+        private readonly IActorRef _consumerActor;
+        private readonly Action<SubSourceStageLogicControl> _subSourceStartedCallback;
+        private readonly Action<(TopicPartition, ISubSourceCancellationStrategy)> _subSourceCancelledCallback;
+        private readonly ISubSourceStageLogicFactory<K, V, TMessage> _subSourceStageLogicFactory;
+        private readonly int _actorNumber;
+        private readonly Decider _decider;
 
-            public Outlet<TMessage> Out { get; }
-            public override SourceShape<TMessage> Shape { get; }
+        public Outlet<TMessage> Out { get; }
+        public override SourceShape<TMessage> Shape { get; }
 
-            public SubSourceStreamStage(TopicPartition topicPartition, IActorRef consumerActor,
-                                  Action<(TopicPartition, IControl)> subSourceStartedCallback,
-                                  Action<(TopicPartition, ISubSourceCancellationStrategy)> subSourceCancelledCallback,
-                                  IMessageBuilder<K, V, TMessage> messageBuilder,
-                                  Decider decider,
-                                  int actorNumber)
-            {
-                _topicPartition = topicPartition;
-                _consumerActor = consumerActor;
-                _subSourceStartedCallback = subSourceStartedCallback;
-                _subSourceCancelledCallback = subSourceCancelledCallback;
-                _messageBuilder = messageBuilder;
-                _decider = decider;
-                _actorNumber = actorNumber;
+        public SubSourceStage(TopicPartition topicPartition, IActorRef consumerActor,
+            Action<SubSourceStageLogicControl> subSourceStartedCallback,
+            Action<(TopicPartition, ISubSourceCancellationStrategy)> subSourceCancelledCallback,
+            Decider decider,
+            int actorNumber, ISubSourceStageLogicFactory<K, V, TMessage> subSourceStageLogicFactory)
+        {
+            _topicPartition = topicPartition;
+            _consumerActor = consumerActor;
+            _subSourceStartedCallback = subSourceStartedCallback;
+            _subSourceCancelledCallback = subSourceCancelledCallback;
+            _decider = decider;
+            _actorNumber = actorNumber;
+            _subSourceStageLogicFactory = subSourceStageLogicFactory;
+
+            Out = new Outlet<TMessage>("out");
+            Shape = new SourceShape<TMessage>(Out);
+        }
 
-                Out = new Outlet<TMessage>("out");
-                Shape = new SourceShape<TMessage>(Out);
-            }
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes)
+        {
+            return _subSourceStageLogicFactory.Create(Shape, _topicPartition, _consumerActor, 
```

Comment:
The `SubSourceStage` is now basically uniform everywhere it gets used - the only thing that really changes is the `ISubSourceStageLogicFactory<K, V, TMessage>` implementation and its generic arguments.